### DISCUSSION TITLE
Add the ability to store CrewAI's memory (short and long) in a database other than Chroma and SQLLite

### DISCUSSION THREAD

**Discussion by RyannosaurusRex** [crewAIInc/crewAI#1125] [Feature Requests]
I'm trying to ship a CrewAI application into production. The current implementation of short and long-term memory use SqlLite and ChromaDB and don't allow the ability to swap out these stores with anything else. 

More specifically, I'd like to use CosmosDB as the store for these, but I don't have a mechanism to write my own implementation. Currently there's not an interface, but the actual methods that store and retrieve data don't seem specific to Chroma/SQLLite at first glance.

I'd really like memory classes like LongTermMemory to be able to accept an overriding implementation instead of CrewAI creating it directly, [like this](https://github.com/crewAIInc/crewAI/blob/f9288295e6093d4b565b8d3a5885482155172dba/src/crewai/memory/long_term/long_term_memory.py#L18C1-L19C1):

`storage = LTMSQLiteStorage()`

I would want the ability to pass my own implementation into the LongTermMemory constructor, and then, if I don't override it with something, it would default to LTMSQLiteStorage to have a great experience out of the box as it currently does today.

TL;DR: 
As someone shipping a CrewAI app to production, I want the ability to write my own implementation for short-term and long-term storage, so that I can store this information externally to support more scaling options as well as to have the option to not *have* to introduce a new database technology within my organization.

**Comment by ryanc4ai** [comment]
This is the discord post related: https://discord.com/channels/1192246288507474000/1250488576404226078

**Comment by slavakurilyak** [comment]
You could use [mem0](https://github.com/mem0ai/mem0) by @mem0ai as it [supports both Qdrant and Chroma](https://docs.mem0.ai/components/vectordb) (vector stores) and [SQLite](https://github.com/mem0ai/mem0/blob/main/mem0/memory/storage.py) (traditional datastore) out of the box

**Comment by indranilr** [comment]
> I'm trying to ship a CrewAI application into production. The current implementation of short and long-term memory use SqlLite and ChromaDB and don't allow the ability to swap out these stores with anything else.
> 
> More specifically, I'd like to use CosmosDB as the store for these, but I don't have a mechanism to write my own implementation. Currently there's not an interface, but the actual methods that store and retrieve data don't seem specific to Chroma/SQLLite at first glance.
> 
> I'd really like memory classes like LongTermMemory to be able to accept an overriding implementation instead of CrewAI creating it directly, [like this](https://github.com/crewAIInc/crewAI/blob/f9288295e6093d4b565b8d3a5885482155172dba/src/crewai/memory/long_term/long_term_memory.py#L18C1-L19C1):
> 
> `storage = LTMSQLiteStorage()`
> 
> I would want the ability to pass my own implementation into the LongTermMemory constructor, and then, if I don't override it with something, it would default to LTMSQLiteStorage to have a great experience out of the box as it currently does today.
> 
> TL;DR: As someone shipping a CrewAI app to production, I want the ability to write my own implementation for short-term and long-term storage, so that I can store this information externally to support more scaling options as well as to have the option to not _have_ to introduce a new database technology within my organization.

Similar ask want to use Postgres as short term and Opensearch as long term memory.
